### ISSUE TITLE
Rework suggestions into a "checklist"

### ISSUE THREAD

**Issue by MartijnHols** [WoWAnalyzer/WoWAnalyzer#895]
This is currently in the idea/prototype phase (although it might move quickly when there are no glaring issues left to tackle). Technical challenges are not yet a major concern.

Current suggestions approach is pretty messy and lacks positive feedback. A checklist would show things you did well (=positive feedback) and all the things left to work on in a structured format. If you come back later having worked on things you should see more checks. If you fucked up you should see less checks and more crosses, but the list and its order is always the same.

The checklist could be grouped by category or majority, or not at all. I thought about grouping it by category, e.g. cast efficiency, but then how/when would we show a positive check if 95% was right but one or two spells were subpar? Maybe it should show a progress bar and be expandable. What determines grouping? Should something like mastery effectiveness be in a general group or its own solo group? This isn't really a major issue I guess.

How do we determine a checklist of things to do? How do we determine it has all important elements? This probably requires a lot of spec knowledge and even then is pretty hard to specify. This is what I thought about for Holy Paladins based on the existing suggestions:

 * Mastery effectiveness
 * Downtime
 * Cast efficiency (casting core spells regularly)
   * Core spells cast efficiency
   * Cooldown cast efficiency
 * Cast behavior (casting the right spells at the right times) (including beacon direct healing)
 * Right talents / talent usage
 * Buffs (legendary buff window usage)
 * Overhealing
 * Others (legendary cap, legendary item level, potions, mana usage)

Would this really be a better format? How would something like `You cast a lot of direct heals on beacon targets. Direct healing beacon targets is inefficient. Try to only cast on beacon targets when they would otherwise die.` be written in a checklist format? Maybe `Don't directly heal beacon targets unnecessarily`, but giving this a good/bad rating makes it seem like we determined if the healing was necessary or not which we can't. Maybe the entire idea just can't work.

**Comment by JacobRReed** [comment]
I feel this is a good idea, however I believe the whole purpose of wowanalyzer is how to improve, not to pat you on the back.

**Comment by MartijnHols** [comment]
@JacobRReed positive feedback when doing something well will makes improving a more pleasant experience, and I think it will help users to keep improving things they're already doing well enough according to our checks.

Adamselene - Today at 00:25
https://docs.google.com/document/d/10aiBVtpZNfvqAqrbVNsm4zq6ayY_cYHAZiexKZK5nx4/edit

**Comment by poneria** [comment]
@JacobRReed @MartijnHols Even disregarding psychological effects, positive feedback is also still useful feedback. It signals that you should keep doing a thing the way you were doing it, instead of also trying to change that, too. 

That said, I think the positive feedback of WoWAnalyzer is already apparent when you get an all-minor or clear list of suggestions. It's exactly like having a clear to-do list, which is a great feeling for me.

@MartijnHols I like the current format of just improvements listed for suggestions. If suggestions are too overwhelming, I usually look closer at the suggested thresholds first to see how tightly the spec module's recommendations are tuned. Otherwise, as a DPS player I usually look for a top 3/top 5 list of things to work on. Maybe the suggestions' toggle can be a tighter slice on the suggestions list -- like a default "top 1-5 improvements to make" in order of importance levels vs a toggle to "all suggestions/verbose mode." The suggestions should help target the things most likely to make a significant improvement, with the option to see all suggestions if power users want that.

As far as visually discerning performance, a variety of results in many analysis categories is usually more indicative of poorer performance, since a good player tends to have high efficiency/low waste across all categories. I don't think a progress bar would be any better, especially since stat boxes/tabs should be already formatted/designed to highlight problem areas.

The stat module boxes on top and the tabs for more graph-using analysis below should already make it apparent how well or not a thing was done. After all, if the analysis stat box/tab doesn't help pinpoint how well an aspect of the spec performance is done, why is it there? I don't see the need for the suggestions section to also duplicate this feature, and the suggestions section would get bogged down trying to do two things at once (1: state what happened; 2: suggest improvements). 

Stat boxes already show things like

- buff uptimes (DoTs)
- proc usage/buffed ability usage (buffs/procs + ability use combined)
- resource usage & gain (usually wasted, but also some general)
- healers: mastery/stat benefit
- pet damage contributions
- downtime default module
- specific ability efficiencies & analysis

in your checklist list, that's:

- [x] Mastery effectiveness
- [x] Downtime
- [x] Cast efficiency (casting core spells regularly)
- [x] Cast behavior (casting the right spells at the right times) (including beacon direct healing)
- [ ] Right talents / talent usage
- [x] Buffs (legendary buff window usage)
- [ ] Overhealing
- [ ] Others (legendary cap, legendary item level, potions, mana usage)

**Talent usage** IDK tech-wise, but you could outline a talent option with a red border and maybe a / ! \ icon when "incorrect" & apply a tooltip to the / ! \ icon + also a suggestion outlining the "correct" talent to use. However, unless the spec page can tell which fight it is for which talent to use, I think the only general suggestion you can make on correct talents used is "hey, you used a lot of X spell, but you're not specced into the favored talent for that."

**Overhealing** could probably be shown with a tab or a stat box. I'm not sure what suggestion you would get out of overhealing that isn't some form of "stop casting when you don't need to" or "wait until a lower health to start casting your heal." (I also don't play a healer main, so maybe I'm being dense.)

**Comment by MartijnHols** [comment]
I worked out more of an example to what the Holy Paladin checklist could look like:

- [ ] Good Mastery usage
      *Mastery has a big impact on the strength of your heals. Try to stay close to the people you are healing to benefit optimally from your Mastery.*
- [ ] Low downtime
      *While it's not optimal to always be casting as a healer, you should still try to always be doing *something* during the entire fight and high downtime is inexcusable. You can reduce your downtime by reducing the delay between casting spells, anticipate movement and move during the GCD, and when you're not healing try to contribute some damage.*
- [ ] Use core spells regularly
      *Spells such as Holy Shock, Light of Dawn and Judgment of Light are our more efficienct spells available. Try to cast them as much as possible (without overhealing).*
- [ ] Use your cooldowns regularly
      *Your cooldowns are an important contributor to your healnig throughput. Try to get in as many casts as you can.*
- [ ] Don't tunnel the tanks
      *A common misconception when it is stated that Holy Paladins are tank healers is that we focus tanks when healing. This is actually inefficient. Let your beacons do the work and ask your co-healers to keep HoTs on tanks and only directly heal the tanks when they would otherwise die.*
- [ ] Don't misuse Light of the Martyr
      *Light of the Martyr is an inefficient spell to cast compared to the alternatives. Try to only cast Light of the Martyr when it will save someone's life or when moving and all other instant cast spells are on cooldown.*
      - [ ] ! You cast {this.inefficientCasts} Light of the Martyrs while Holy Shock was available. Try to **never** cast Light of the Martyr when something else is available.

**Comment by MartijnHols** [comment]
@poneria You have some valid points and I think a rework of the suggestions structure can be done with all of that still in mind. 

My primary goal is to reduce the clutter in reports like the ones suggested as examples on the homepage (reminder: please volunteer your own logs as examples, I'd like some newer/better examples up there).
https://wowanalyzer.com/report/LRchdHVAmWtNQ8Fj/22-Mythic+Harjatan+-+Kill+(5:54)/Zerotorescue

![image](https://user-images.githubusercontent.com/4565223/33516652-b404cbf8-d776-11e7-8e1c-d990511406cf.png)

Most of the suggestion are fair, and looking at each suggestion individually their shown importance is pretty accurate too. Maybe HS should be major while Tyr's which has less of an impact could be just average but that's nitpicking.

My issue with it is that because of the amount it's hard to get something actionable. This might be due to a personal quirk of mine where I have a much harder time deciding what to do when I have a lot of things to do versus when my todo list is short with just a few things. That might not even be just a personal quirk when I consider there are systems that actively combat this problem (e.g. Kanban boards restrict the amount of tasks in a list).

I can think of two ways to reduce the amount of items in a list: by not showing less important things or by combining very small and similar tasks into one bigger task. We're doing the first one by hiding minor importance and I don't think it's enough. As seen in the list above there are still a lot of things left making it hard to pick something to work on.

This is why I am considering the approach of making a predefined checklist with a limited amount of top-level items. The example in my previous comment is a large part of what I think it would look like for Holy Paladins. Clicking an item expands it, showing the suggestions underneath it, e.g. "Use core spells regularly" would show the cast efficiency suggestions for Holy Shock and Light of Dawn at the least.

I think that when we have such a limited list of to-level things to work on it would also make sense to show them all so that we can show users positive feedback when they did the things well. I think this would achieve that separate goal nicely. It is true that the statistic boxes also reveal most of this information, but not everyone will (take the time to) understand them and they're also pretty cluttered and daunting. I think the statistic boxes are good for people with good understanding of what we're analyzing and what it correct behavior, but pretty useless for anyone else. (some exceptions such as item performances)

The progress bars would be indicators of how well a checklist item was done. E.g. at 60% actual/80% recommended cast efficiency it could be 75% full. I think this would be better than the vague minor/average/major.

**Comment by MartijnHols** [comment]
Suggestions -> Checklist items
https://docs.google.com/spreadsheets/d/1D3lHWhh22QQuea48T8aSjxPSsnpS--jE18Gy6GbgVZ0/edit?usp=sharing

**Comment by poneria** [comment]
@MartijnHols 

> This might be due to a personal quirk of mine where I have a much harder time deciding what to do when I have a lot of things to do versus when my todo list is short with just a few things. That might not even be just a personal quirk when I consider there are systems that actively combat this problem (e.g. Kanban boards restrict the amount of tasks in a list).

That was my idea, basically -- to restrict the number of suggestions/actionables they see, period, with the option to extend if the power user wants to see everything. There's a small list of actionables that will likely have a major/core impact, or minor/circumstance if the user has mastered the core rotation. 

If the user has more than 5 major/core things to work on, it still only shows 5 for them to work on, because when they come back next raid, hopefully those 5 will be improved specifically & there will be 5 more issues to work on, rather than attempting to choose or improve all the items on an overloaded progress bar list. Order of a top 5 in a situation of 5+ major items will have to be determined by the spec maintainer; I figure you answer the question in the style of "if you could only remember 1 thing from all this, it would be X." On the one hand, that's a hard question; on the other hand, welcome to guide-writing :) . 

Although a progress bar can show you positive feedback as you grow, the list of actionables under it are still going to be a daunting mental load to work on if they have lots of things "wrong." I realize you're chunking the suggestions, but once a user with a lot of problems opens one or more of the chunks, it's still an overloaded list of things to get done. Let me know if I'm visualizing this wrong, but moving the suggestions around into partially hidden interfaces (albeit with fun/pretty progress bars) is what's being accomplished, rather than actually paring down the display overload.

---
Separately, but related: different positive feedback idea -- if there was a summary status of the suggestions, like a general comment on how well you did, but phrased positively/looking to improvement rather than back at issues. Examples:

| Problem types | Message |
| --------------- | --------- | 
| 5+ Major | You have a lot to work on, but we've got your back. Let's start with these five things. |
| Major/Average mix | Good job! Let's work on bringing these core items up to speed. |
| Average/Minor mix | Excellent! Let's work on enhancing your rotation. |
| Minor only | Awesome! Let's make it perfect! |
| Zero problems | Amazing! Chase those parsing ranks!

**Comment by MartijnHols** [comment]
I think just limiting the amount of shown suggestions would be unpredictable and confusing for users and not be enough gain. I dislike information being shown / hidden based on an arbitrary rules as this is unpredictable and I believe interfaces should work just as you'd expect without having any additional knowledge. It could also be confusing to think the shown things are the only things wrong and when you get a few new ones the next visit while old ones disappear remove any form of progression.

I've also begun to see the checklist as a good way to get spec maintainers to really think about what defines a good result. I hadn't really thought about this for Holy Paladins until I made the checklist example above. I just randomly made suggestions as I learned about issues I could detect. I think starting with a checklist would have made me think about the importance of things and get me more focused on what's actually important. This is just an additional benefit I wanted to mention, one that I think is also very important (not because we really need it, but because I think it will ease development and improve the end result).

`Although a progress bar can show you positive feedback as you grow, the list of actionables under it are still going to be a daunting mental load to work on if they have lots of things "wrong." I realize you're chunking the suggestions, but once a user with a lot of problems opens one or more of the chunks, it's still an overloaded list of things to get done.`

Actually an advantage of checklist items is that things that can be solved together can be grouped. The behavioral change required to pass "Use core spells on cooldown" is to focus on keeping spells on cooldown more. I do not think it's a lot harder to try to do this for all relevant spells rather than just one.

Funny enough I started working on a feature at work this week that is almost identical to this issue. There we decided to go with "Categories" as top-level items and "Requirements" as sub-items with each "requirement" being ticked or crossed based on the result and if >=1 requirement is wrong the entire category is marked wrong. We might not be able to transform our suggestions into a good requirement format so our solution might not be the same.

I'm pretty convinced a more checklist-like format is going to be a big improvement. I made a mockup that I've already begun working on:

![New results page mockup](https://cdn.discordapp.com/attachments/316864385052049420/386839276932956161/new-results.png)

**Comment by Sharrq** [comment]
Im not sure if your mockup is exactly how you plan on making the checklist look or not, but i think it might be beneficial to give more detail than just the category i.e. "Cast Efficiency" or "Cooldown Use".

Im trying to find a website that has roughly what i was imagining in my head, but this kinda shows it halfway decently (obviously with not nearly as much content as in the image, but it gives the general idea of what i was imagining ... where you have your main heading followed by what is included in that heading
https://d2v4zi8pl64nxt.cloudfront.net/launching-a-new-website-your-seo-checklist-whiteboard-friday/58f048d2b0ea30.57947942.jpg

**Comment by MartijnHols** [comment]
The checklist part is just reserved space. How it will look like is still a mystery.

**Comment by Sharrq** [comment]
You could probably use the same boxes that the suggestions use currently and just make them narrower and maybe a little shorter. Without making a Frost Mage checklist, i would imagine most specs could fit everything into 10-12 summary items at most which should be enough to fill that space. 

Using the suggestions as a guide, here is what it looks like next to the Holy Pally Stat Tracker/Items/etc with 8 items, but you could likely fit at least 10 there with no issue, especially if you made the suggestion boxes a little bit shorter.
https://imgur.com/a/jBsvl

with the amount of blank space, i would think you could easily convert those to a Summary/Description type layout with a spell icon on the left side. or replace the spell icon with an oversized checkbox or something like that.... would probably look pretty nice.

**Comment by MartijnHols** [comment]
I'm currently thinking it would be defined something like this:

![image](https://user-images.githubusercontent.com/4565223/33785575-f27e7cdc-dc65-11e7-8314-ed4f80145264.png)

It would be something like "Rules" are the behavioral things e.g. `Use core spells on cooldown`, `Try to avoid a high amount of downtime`, `Only use Light of the Martyr when absolutely necessary`, `Position yourself well to utilize your Mastery` (naming them is pretty hard), and "Requirements" would be simple things that are being checked like specific spells and they would have a gauge like a color-coded progress bar (like Cast Efficiency's efficiency bar) to indicate how well one did on that requirement. I think the rule's effectiveness would match the lowest effectiveness (although perhaps that could be configurable per rule).

I think this approach makes the checklist pretty simple to use; the main points are what matter, expanding it (since they start collapsed) show a detailed view of what was measured and determined to be sufficient or insufficient.

I worked out what I think it should be for Holy Paladins in a spreadsheet: https://docs.google.com/spreadsheets/d/1D3lHWhh22QQuea48T8aSjxPSsnpS--jE18Gy6GbgVZ0/edit#gid=1742405434

**Comment by MartijnHols** [comment]
# The Checklist

> The checklist is for most users the most important feature of the application. With good rules, explanations and checks it will point out gameplay issues that they can then work on to improve their performance. Working this out well will help a lot of people.

The Checklist shows a list of "rules" that one should follow to play optimally. These rules are mostly non-technical and with enough class knowledge and giving it enough thought you should be able to prepare a list of rules for the checklist.
![Rules overview](https://user-images.githubusercontent.com/4565223/33906368-057712ee-df82-11e7-8348-962623cb53bb.png)

Within a rule is a description to explain the rule and its possible exceptions to the user and show the things that are relevant and being checked for the rule (called "Requirements"). We show *what* is being measured, the *result* and a *performance bar*. The result gives the user a quick view of how/what he did without having to search for the metric elsewhere in the analyzer, and the performance simplifies it to a "bad-good" rating.
![Expanded rule](https://user-images.githubusercontent.com/4565223/33906410-22059304-df82-11e7-8a18-10c66945d15a.png)

The *requirement* settings are based around suggestion thresholds.

# Preparing a new checklist

> This step doesn't really involve any code.

> TL;DR: Figure out a complete checklist of specific and actionable rules before writing any code.

It's not easy to figure out a good checklist and it may take some time to get a complete set of rules before even writing a single line of code. The example above was the initial list for Holy Paladins, but before we got there it went through several iterations. I'll explain the process I went through.

I started with a basic list of important things separated to a point that the user can fix individually:

 * Mastery effectiveness
 * Downtime
 * Cast efficiency (casting core spells regularly)
   * Core spells cast efficiency
   * Cooldown cast efficiency
 * Cast behavior (casting the right spells at the right times) (including beacon direct healing)
 * Right talents / talent usage
 * Buffs (legendary buff window usage)
 * Overhealing
 * Others (legendary cap, legendary item level, potions, mana usage)

I tried to make it complete without going too much in-depth. This list of *rules* formed a good basis for working out a more in-depth list:

- [ ] Use core spells on cooldown
      *Spells such as Holy Shock, Light of Dawn and Judgment of Light are our more efficient spells available. Try to cast them as much as possible (without overhealing).*
- [ ] Use your cooldowns regularly
      *Your cooldowns are an important contributor to your healing throughput. Try to get in as many casts as you can.*
- [ ] Use other spells regularly
      *Some spells have a small impact on your performance but will help make fights easier. Try to use them effectively so long as it doesn't degrade other things in this list.*
- [ ] Good positioning / good Mastery effectiveness
      *Mastery has a big impact on the strength of your heals. Try to stay close to the people you are healing to benefit optimally from your Mastery.*
- [ ] Low downtime
      *While it's not optimal to always be casting as a healer, you should still try to always be doing *something* during the entire fight and high downtime is inexcusable. You can reduce your downtime by reducing the delay between casting spells, anticipate movement and move during the GCD, and when you're not healing try to contribute some damage.*
- [ ] Don't tunnel the tanks
      *A common misconception when it is stated that Holy Paladins are tank healers is that we focus tanks when healing. This is actually inefficient. Let your beacons do the work and ask your co-healers to keep HoTs on tanks and only directly heal the tanks when they would otherwise die.*
- [ ] Don't misuse Light of the Martyr
      *Light of the Martyr is an inefficient spell to cast compared to the alternatives. Try to only cast Light of the Martyr when it will save someone's life or when moving and all other instant cast spells are on cooldown.*
      subitem (if this is true this item could get a massively higher importance): - [ ] ! You cast {this.inefficientCasts} Light of the Martyrs while Holy Shock was available. Try to **never** cast Light of the Martyr when something else is available.

I changed the order to show more important and specific things at the top, I reworded a bunch of things to sound more like *things to do* and experimented with different wording for *mastery effectiveness*, and downtime. "Cast behavior" was split into two rules that would be easier for users to follow; "Don't tunnel the tanks" and "Don't misuse Light of the Martyr". This is a lot more specific and actionable. I also left a few less important things out, such as picking the right talents, overhealing and "others". I felt these weren't important enough to delay the main things for.

Adding the descriptions helped me justify the rules. I found that something you can't explain well is often something that's wrong, and each rule and its description should only be about *one* thing. If this fails consider changing your rules.

With all of the previous information I created a list of the rules and requirements I wanted to add in Google Docs: https://docs.google.com/spreadsheets/d/1D3lHWhh22QQuea48T8aSjxPSsnpS--jE18Gy6GbgVZ0/edit#gid=1742405434. This is a simple overview without any unnecessary information allowing for quick feedback (from other experts of my spec) and iteration. When I was satisfied with the result I recreated it in the code.

## Existing spec

One thing you can do to help yourself figure out the checklist is to make a list of all suggestions already in place and to try to group those to checklist items. I did this here: https://docs.google.com/spreadsheets/d/1D3lHWhh22QQuea48T8aSjxPSsnpS--jE18Gy6GbgVZ0/edit#gid=0. Given enough existing suggestions, this gives a fairly complete picture of what the checklist should at least include.

## New spec

I think if you make a full checklist concept before starting with anything else it will give you a clear focus on the most important things for your spec.

# Making the checklist

> We start writing code now.

**Comment by MartijnHols** [comment]
Implemented
